### ISSUE TITLE
using hono requires skipLibCheck

### ISSUE THREAD

**Issue by aroman** [honojs/hono#1200]
Hi,

I've just tried adding your library to my project. However, simply running `new Hono();` breaks my typescript project due to numerous type issues with your bundled types; some of your types expect the global `Request` to be a DOM request object, but this is not true in the workers environment.

Passing `skipLibCheck: true` to my tsconfig suppresses the errors, which I notice is what the [hono repo itself does in its own tsconfig](https://github.com/honojs/hono/blob/4b2dfb824ba68c6dd03ea93ba60649a95d370e80/tsconfig.json#L10).

However, this is definitely not a good thing to do, as it disables ALL type checking for all dependencies.

Here is a demonstration. Any ideas? I've also included my tsconfig below. 

```
~/P/server main• 1s ❱ tsc --noEmit --project .
~/P/server main• ❱ echo "Note: no type errors in my project prior to adding hono"
Note: no type errors in my project prior to adding hono
~/P/server main• ❱ npm install hono

added 1 package, and audited 915 packages in 883ms

152 packages are looking for funding
  run `npm fund` for details

1 moderate severity vulnerability

To address all issues, run:
  npm audit fix

Run `npm audit` for details.
~/P/server main• 1s ❱ echo -e "import { Hono } from 'hono';\n const app = new Hono();" >> src/worker.ts
~/P/server main• ❱ tsc --noEmit --project .
../node_modules/hono/dist/types/client/types.d.ts:22:18 - error TS2430: Interface 'ClientResponse<T>' incorrectly extends interface 'Response'.
  The types returned by 'json()' are incompatible between these types.
    Type 'Promise<T>' is not assignable to type 'Promise<T>'. Two different types with this name exist, but they are unrelated.
      Type 'T' is not assignable to type 'T'. Two different types with this name exist, but they are unrelated.
        'T' could be instantiated with an arbitrary type which could be unrelated to 'T'.

22 export interface ClientResponse<T> extends Response {
                    ~~~~~~~~~~~~~~

  ../node_modules/hono/dist/types/client/types.d.ts:22:33
    22 export interface ClientResponse<T> extends Response {
                                       ~
    This type parameter might need an `extends T` constraint.

../node_modules/hono/dist/types/request.d.ts:51:21 - error TS2304: Cannot find name 'RequestRedirect'.

51     get redirect(): RequestRedirect;
                       ~~~~~~~~~~~~~~~

../node_modules/hono/dist/types/request.d.ts:54:18 - error TS2304: Cannot find name 'RequestCache'.

54     get cache(): RequestCache;
                    ~~~~~~~~~~~~

../node_modules/hono/dist/types/request.d.ts:55:24 - error TS2304: Cannot find name 'RequestCredentials'.

55     get credentials(): RequestCredentials;
                          ~~~~~~~~~~~~~~~~~~

../node_modules/hono/dist/types/request.d.ts:58:17 - error TS2304: Cannot find name 'RequestMode'.

58     get mode(): RequestMode;
                   ~~~~~~~~~~~

../node_modules/hono/dist/types/request.d.ts:60:26 - error TS2304: Cannot find name 'ReferrerPolicy'.

60     get refererPolicy(): ReferrerPolicy;
                            ~~~~~~~~~~~~~~


Found 6 errors in 2 files.

Errors  Files
     1  ../node_modules/hono/dist/types/client/types.d.ts:22
     5  ../node_modules/hono/dist/types/request.d.ts:51
~/P/server main• | 2 ❱
```

```tsconfig
...
  "compilerOptions": {
    "lib": ["es2022"],
    "types": ["@cloudflare/workers-types"]
  },
...
  ```

**Comment by yusukebe** [comment]
Hi @aroman !

Thank you for raising the issue. As you said, `skipLibCheck` should be `false` and I've fixed the issue in #1201 .

**Comment by aroman** [comment]
thanks @yusukebe for the quick followup! 

while i'm glad this project is now not using `skipLibCheck`, unfortunately I believe that using hono would still require _my project_ (or anyone else targetting Cloudflare) to use `skipLibCheck` to use hono.

As I mentioned before, your types e.g. in `[request.d.ts](https://github.com/honojs/hono/blob/main/src/request.ts)` assume the `Request` is a browser (DOM) request object, which it will not be when running in e.g. Cloudflare. 

These are the errors I get when using hono, even with a version I built locally and linked (`npm run build && npm link`) including #1201. Note the missing types for `RequestCache`, `RequestRedirect` etc. [@cloudflare/workers-types](https://www.npmjs.com/package/@cloudflare/workers-types/v/4.20230518.0?activeTab=code) does not define these types, so they will result in type errors.

```
npm run typecheck

> typecheck
> tsc --noEmit --project client && tsc --noEmit --project server

../hono/dist/types/client/types.d.ts:22:18 - error TS2430: Interface 'ClientResponse<T>' incorrectly extends interface 'Response'.
  The types returned by 'json()' are incompatible between these types.
    Type 'Promise<T>' is not assignable to type 'Promise<T>'. Two different types with this name exist, but they are unrelated.
      Type 'T' is not assignable to type 'T'. Two different types with this name exist, but they are unrelated.
        'T' could be instantiated with an arbitrary type which could be unrelated to 'T'.

22 export interface ClientResponse<T> extends Response {
                    ~~~~~~~~~~~~~~

  ../hono/dist/types/client/types.d.ts:22:33
    22 export interface ClientResponse<T> extends Response {
                                       ~
    This type parameter might need an `extends T` constraint.

../hono/dist/types/hono-base.d.ts:1:25 - error TS2307: Cannot find module './context' or its corresponding type declarations.

1 import { Context } from './context';
                          ~~~~~~~~~~~

../hono/dist/types/hono-base.d.ts:2:39 - error TS2307: Cannot find module './context' or its corresponding type declarations.

2 import type { ExecutionContext } from './context';
                                        ~~~~~~~~~~~

../hono/dist/types/index.d.ts:3:50 - error TS2307: Cannot find module './context' or its corresponding type declarations.

3 export type { Context, ContextVariableMap } from './context';
                                                   ~~~~~~~~~~~

../hono/dist/types/request.d.ts:51:21 - error TS2304: Cannot find name 'RequestRedirect'.

51     get redirect(): RequestRedirect;
                       ~~~~~~~~~~~~~~~

../hono/dist/types/request.d.ts:54:18 - error TS2304: Cannot find name 'RequestCache'.

54     get cache(): RequestCache;
                    ~~~~~~~~~~~~

../hono/dist/types/request.d.ts:55:24 - error TS2304: Cannot find name 'RequestCredentials'.

55     get credentials(): RequestCredentials;
                          ~~~~~~~~~~~~~~~~~~

../hono/dist/types/request.d.ts:58:17 - error TS2304: Cannot find name 'RequestMode'.

58     get mode(): RequestMode;
                   ~~~~~~~~~~~

../hono/dist/types/request.d.ts:60:26 - error TS2304: Cannot find name 'ReferrerPolicy'.

60     get refererPolicy(): ReferrerPolicy;
                            ~~~~~~~~~~~~~~

../hono/dist/types/types.d.ts:1:30 - error TS2307: Cannot find module './context' or its corresponding type declarations.

1 import type { Context } from './context';
                               ~~~~~~~~~~~


Found 10 errors in 5 files.

Errors  Files
     1  ../hono/dist/types/client/types.d.ts:22
     2  ../hono/dist/types/hono-base.d.ts:1
     1  ../hono/dist/types/index.d.ts:3
     5  ../hono/dist/types/request.d.ts:51
     1  ../hono/dist/types/types.d.ts:1
```

**Comment by yusukebe** [comment]
Hi @aroman !

Thanks for the comment. I've created another PR #1206 to fix it. Could you review it?

**Comment by aroman** [comment]
Nice! I tried your branch locally, and I can confirm it fixes all of the type errors I was having with cloudflare types 😄 

I am not really familiar with `hono` (this issue was preventing me from using it), so I while I took a quick look at #1206, I am not qualified to review it. I did leave one comment/question. You might want to test more thoroughly to confirm it doesn't break things in other environments like node/deno/etc.

Thank you again!
